sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PullOutNestedDataOuterRefExpressions.scala

```diff
@@ -0,0 +1,133 @@
+ * other object.
+ *
+ * Example: Here, we have outer table x and inner table y in a scalar subquery, correlated
+ * on xm[1] = ym[1] where xm and ym are map columns.
+ *
+ * The plan before the rewrite is:
+ *
+ * Filter (scalar-subquery#50 [xm#11] > cast(2 as bigint))
+ *  +- Aggregate [sum(y2#14) AS sum(y2)#52L]
+ *     +- Filter (outer(xm#11)[1] = ym#13[1])
+ *        +- Relation spark_catalog.default.y[ym#13,y2#14] parquet
+ *  +- Relation spark_catalog.default.x[xm#11,x2#12] parquet
+ *
+ * The plan after the rewrite adds a projection for xm[1] to the outer plan, and replaces the outer
+ * reference inside the subquery with that:
+ *
+ * Project [xm#11, x2#12]
+ * +- Filter (scalar-subquery#50 [xm[1]#55] > cast(2 as bigint))
+ *    :  +- Aggregate [sum(y2#14) AS sum(y2)#52L]
+ *    :     +- Filter (outer(xm[1]#55) = ym#13[1])
+ *    :        +- Relation spark_catalog.default.y[ym#13,y2#14] parquet
+ *    +- Project [xm#11, x2#12, xm#11[1] AS xm[1]#55]
+ *       +- Relation spark_catalog.default.x[xm#11,x2#12] parquet
+ *
+ * This is implemented as a separate rule from DecorrelateInnerQuery because it's much simpler
+ * and safer, and also benefits us when the same nested data expression is used multiple times.
+ * In particular:
+ * - In DecorrelateInnerQuery, outer references is an AttributeSet, so it can't store general
+ *   expressions. In principle we could change this but it would add substantial complexity.
+ * - DecorrelateInnerQuery only manipulates the inner query, not the outer plan, whereas
+ *   this rewrite needs to add projections to the outer plan
+ */
+object PullOutNestedDataOuterRefExpressions extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (SQLConf.get.getConf(SQLConf.PULL_OUT_NESTED_DATA_OUTER_REF_EXPRESSIONS_ENABLED)) {
+      rewrite(plan)
+    } else {
+      plan
+    }
+  }
+
+  def rewrite(plan: LogicalPlan): LogicalPlan = plan.transformUpWithPruning(
+    _.containsAllPatterns(PLAN_EXPRESSION, OUTER_REFERENCE, EXTRACT_VALUE)) {
+    case plan: UnaryNode =>
+      // Map of original expression semanticHash to (new outer projection, attribute reference)
+      val newExprMap = mutable.HashMap.empty[Int, (NamedExpression, AttributeReference)]
+      val newPlan = plan.transformExpressionsWithPruning(
+        _.containsAllPatterns(PLAN_EXPRESSION, OUTER_REFERENCE, EXTRACT_VALUE)) {
+        case e: DynamicPruningSubquery => e // Skip this case
+        case subqueryExpression: SubqueryExpression =>
+          val innerPlan = subqueryExpression.plan
+          val newInnerPlan = innerPlan.transformAllExpressionsWithPruning(
+            _.containsAllPatterns(OUTER_REFERENCE, EXTRACT_VALUE)) {
+            // We plan to extend to other ExtractValue and similar exprs in future PRs
+            case e @ GetMapValue(outerRef: OuterReference, key) =>
+              val hash = e.semanticHash()
```

Comment:
what's the significance of semantic hash rather than the regular hash here?

Reply:
It would allow us to reuse the same expression when there are cosmetic differences (e.g. different ordering for commutative operations)

Reply:
Added a test case `select * from x where (select sum(y2) from y where xm[x2+1] = ym[1] and xm[1+x2] = ym[2]) > 2; -- Two key expressions that are semantically equal`